### PR TITLE
Add new synthesis recipes to API.

### PR DESCRIPTION
**Pull request by hhaoyan**
Adding new schema for synthesis recipes as we want to incorporate the latest solid-state/sol-gel synthesis datasets at https://github.com/CederGroupHub/text-mined-synthesis_public.

## Contributor Checklist

- [x] I have broken down my PR scope into the following TODO tasks
   -  [x] Define new pydantic models.
   -  [x] Write queries. The following queries will be implemented:
      - [x] Query by target/precursor material composition.
      - [x] Query by synthesis type.
      - [x] Query by paragraph string (keyword matching).
      - [x] Filtering by experimental operations.
      - [x] Filtering by experimental conditions.
   -  [x] Implement script that ports the public dataset from [here](https://github.com/CederGroupHub/text-mined-synthesis_public) to MP database.
- [x] ~~I have run the tests locally and they passed.~~
- [x] ~~I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR~~

### CODE REVIEW

**Comment by hhaoyan** [comment]
Hi @mkhorton could you elaborate on what tests should be implemented for this particular API? Thanks!

**Comment by mkhorton** [comment]
Right now, no tests (shocking, I know!). We're just building up our test suite for this repo and figuring out what they should look like, starting with client tests, maybe @munrojm can comment.

In any case, this can be merged without tests if it's otherwise good.

**Comment by munrojm** [comment]
Hi @hhaoyan, this looks great so far! Thanks for putting it together. As @mkhorton said, I am in the process of getting a full initial suite of tests implemented, so don't worry about that portion. 

Let us know when you feel it is ready to merge.

File: src/mp_api/routes/synthesis/models/core.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    # Basic facts about this recipe:
+    doi: str = Field(
+        ...,
+        description="DOI of the journal article.",
+    )
+    paragraph_string: str = Field(
+        "",
+        description="The paragraph from which this recipe is extracted."
+    )
+    synthesis_type: str = Field(
```
Comment (mkhorton) [inline src/mp_api/routes/synthesis/models/core.py:25]: Can this be an enum? If there are only a few types of synthesis?

**Comment by hhaoyan** [comment]
@mkhorton @munrojm I'm thinking to do some post processing in Query classes based on the query parameters. However, the current function `def post_process(self, docs: List[Dict])` only accepts a single parameter `docs`. Is there a proper way to pass the query parameters to `post_process`?

**Comment by munrojm** [comment]
@hhaoyan, not currently. What sort of post-processing do you have in mind? Right now the idea is to have the `post_process` method run for both the search and key name endpoints. This would not support using values of the query parameters that are only seen in the former.

**Comment by hhaoyan** [comment]
I’m trying to selectively return part of a full text paragraph based on search keywords, which requires the post processor to have access to the search keywords. For example,if a search keyword is preset in the second sentence, then only that sentence will be returned. This is somewhat similar to what google search results look like. Is there a way I can achieve this?

**Comment by mkhorton** [comment]
Yes, this is how I've done the highlighting on the current synthesis data -- I just set `maxNumPassages` to 1 and I believe that's below the 150 limit. Whether we post-process or not, that highlighting functionality is definitely what we want to use. It's fast and we've already written the frontend to handle the output.

**Comment by hhaoyan** [comment]
Thanks for the info @mkhorton @munrojm ! I looked at what had been done in `/text_search/` and I think this should be enough. However I don't have access to a MongoDB Atlas subscription and couldn't test this feature on my local computer. If that's possible, I'd like to have some access to a test environment so I can test & convert all the data records from our side.

**Comment by hhaoyan** [comment]
@mkhorton I think this could be merged now!

**Comment by munrojm** [comment]
@hhaoyan, can you confirm that you are able to query properly when running the API locally?

**Comment by hhaoyan** [comment]
Just had a conversation with @codytodonnell and it seemed to work well.

**Comment by munrojm** [comment]
@hhaoyan, okay great. Do you mind fixing the mypy linting issues before I merge? Don't worry about the second set of tests.

**Comment by hhaoyan** [comment]
OK sure! 

Also, the code has 9 `ensure_index` function calls when each query is run . Do you think this would slow down the API? Do you have any suggestions on how to make `ensure_index` to be run just by once?

The code that runs 9 `ensure_index` is here:

https://github.com/materialsproject/api/pull/257/files#diff-03cb6bf4cb52b7cc45905fd54f7cb1e29314ae92bc32da135f78b526059d2e98R209

**Comment by munrojm** [comment]
I am going to merge this in with another branch that should allow me to define a separate query operator for this custom function. That should allow it to only run when in debug mode. I will take care of that change after this is merged. 

For now, can you simply comment out that code block?

**Comment by mkhorton** [comment]
This looks great, my only comment would be to add a docstring to the data_adaptor files to clarify what they do, what data they take, etc. and for some of the functions in those files too.

**Comment by hhaoyan** [comment]
OK, the mypy lint was fixed and I added docstrings to data adaptors. 

I also commented the  block of ensuring indexes.

**Comment by munrojm** [comment]
Great, thanks @hhaoyan!

**Comment by mkhorton** [comment]
Thanks Haoyan!

**Comment by hhaoyan** [comment]
Thanks @mkhorton and @munrojm !

**Pull request merged**